libdevice/device_itt.h

```diff
@@ -0,0 +1,58 @@
+//==------- device_itt.h - ITT devicelib functions declarations ------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//==------------------------------------------------------------------------==//
+
+#ifndef __LIBDEVICE_DEVICE_ITT_H__
+#define __LIBDEVICE_DEVICE_ITT_H__
+
+#include "device.h"
+
+#ifdef __SPIR__
+#include <cstddef>
+#include <cstdint>
+
+// Use SPIRV constants directly in place of OCL intrinsic functions.
+#define __SPIRV_VAR_QUALIFIERS EXTERN_C const
+typedef size_t size_t_vec __attribute__((ext_vector_type(3)));
+__SPIRV_VAR_QUALIFIERS size_t __spirv_BuiltInGlobalLinearId;
+__SPIRV_VAR_QUALIFIERS size_t_vec __spirv_BuiltInWorkgroupId;
+__SPIRV_VAR_QUALIFIERS size_t_vec __spirv_BuiltInWorkgroupSize;
+
+#define ITT_STUB_ATTRIBUTES __attribute__((noinline, optnone))
+
+// FIXME: must be enabled via -fdeclare-spirv-builtins
+DEVICE_EXTERN_C char __spirv_SpecConstant(int, char);
+
+#define ITT_SPEC_CONSTANT 0xFF747469
+
+static inline bool isITTEnabled() {
+  return __spirv_SpecConstant(ITT_SPEC_CONSTANT, 0) != 0;
+}
+
+// Wrapper APIs that may be called by compiler-generated code.
+DEVICE_EXTERN_C
+void __itt_spirv_wi_start_wrapper();
+DEVICE_EXTERN_C
+void __itt_spirv_wi_finish_wrapper();
+DEVICE_EXTERN_C
+void __itt_spirv_wg_barrier_wrapper();
+DEVICE_EXTERN_C
+void __itt_spirv_wi_resume_wrapper();
+
+// Non-inlinable and non-optimizable APIs that are recognized
+// by profiling tools.
+DEVICE_EXTERN_C ITT_STUB_ATTRIBUTES void
+__itt_spirv_wi_start_stub(size_t *group_id, size_t wi_id, uint32_t wg_size);
+DEVICE_EXTERN_C ITT_STUB_ATTRIBUTES void
+__itt_spirv_wi_finish_stub(size_t *group_id, size_t wi_id);
+DEVICE_EXTERN_C ITT_STUB_ATTRIBUTES void
+__itt_spirv_wg_barrier_stub(uintptr_t barrier_id);
+DEVICE_EXTERN_C ITT_STUB_ATTRIBUTES void
+__itt_spirv_wi_resume_stub(size_t *group_id, size_t wi_id);
```

Comment:
If there stubs are the ones to be replaced by analysis tool, could we use names that we discussed previously?
void __itt_offload_wi_start(size_t* group_id, size_t wi_id, uint32_t wg_size);
void __itt_offload_wi_finish(size_t* group_id, size_t wi_id);
void __itt_offload_wg_barrier(uintptr_t barrier_id);
void __itt_offload_wi_resume(size_t* group_id, size_t wi_id);

What about other stubs?
void __itt_offload_sync_acquired(uintptr_t sync_id);
void __itt_offload_sync_releasing(uintptr_t sync_id);

void __itt_offload_wg_local_range(void* ptr, size_t size);

void __itt_offload_atomic_op_start(void* object, __itt_atomic_mem_op_t op_type, __itt_atomic_mem_order_t mem_order);
void __itt_offload_atomic_op_finish(void* object, __itt_atomic_mem_op_t op_type, __itt_atomic_mem_order_t mem_order);

Reply:
Right, I changed the naming scheme while implementing it, because I saw that we have different names in different discussions.  For example, I think, we agreed on `__offload_itt` prefix for the `stubs`, but then you used `__itt_offload` in `offload_annotations.h`.  Moreover, I see `__itt_sync_acquired` without `offload` in `offload_annotations.h`.  It looked like a mess to me, so I decided to use a "cleaner" naming scheme:
* Use `spirv` suffix/prefix for this is a SPIR-V device library.
* User APIs: do not use any new suffix/prefix, e.g. `__itt_sync_acquired`, `__itt_sync_releasing`.
* Stubs: use `__itt_spirv_` prefix and `_stub` suffix.
* Compiler wrappers: use `__itt_spirv_` prefix and `_wrapper` suffix.

`offload` prefix/suffix seems too generic to me now.  Since we are using `wi`, `wg` names, we have to make it clear that these are `SPIR-V` device only APIs.

Please list the names you expect to see for the `stubs` (the APIs recognized by the tools), and for the `user APIs`.  Let's take `wi_start` and `sync_acquired` as an example.

Reply:
Yes, we used different names in the prototype. But then I explicitly use ittnotify naming convention in all documents: header file, slides and the document. Honestly, I would not change those names there once again since it is used by different people now. If you insist on changing names, then we can surely do that.

Vlad had suggestion to split all annotations to the ones that are valid for offload code and the ones applicable for ‘native’ code. That’s the reason you see __itt_sync_acquired() with- and without- ‘offload’ prefix.

All functions in offload_annotations.h are expected to be visible to applications and runtime libraries. They should have "clean" look. Internally we can use any prefix or suffix for ‘stub’ functions. I will be fine with:
__itt_offload_wi_start()          - public API
__itt_offload_wi_start_stub() - internal API

Reply:
My point is that we cannot have the same names for user APIs and the APIs that are recognized by the tools (Inspector).  So if the APIs in `offload_annotations.h` are user APIs, then I will have to add `_stub` suffix, and the Inspector will have to recognize APIs with `_stub` suffix.  Do you agree?

I missed the fact that there is `__itt_sync_acquired` with and without `offload` prefix.  Thanks!

Again, I think `wi`/`wg` and the comments in `offload_annotations.h` are quite specific to `SPIR-V` offload, so it seems to me we'd rather make it clear in the API names.  But I will stick to `__itt_offload`.

Reply:
> My point is that we cannot have the same names for user APIs and the APIs that are recognized by the tools (Inspector). So if the APIs in `offload_annotations.h` are user APIs, then I will have to add `_stub` suffix, and the Inspector will have to recognize APIs with `_stub` suffix. Do you agree?

Yes, this will work for me.